File: Point.cpp
#include <sstream>
#include <string>
#include <iostream>
#include "Point.h"

//Point::Point() : m_x{ 0.0 }, m_y{ 0.0 }
//{
//    std::cout << ">Point::Point()" << std::endl;
//    // передаём инициализацию объекта другому методу
//    read();
//}
//
//Point::Point(double x, double y) : m_x{ x }, m_y{ y }
//{
//    std::cout << ">Point::Point(" << x << ", " << y << ")" << std::endl;
//}

Point::Point()
{
    // если конструктор написать так, то для каждого скрытого поля
    // сначала вызовется КОНСТРУКТОР ПО УМОЛЧАНИЮ (для double = 0.0)
    // и только затем произойдёт присваивание
    std::cout << ">Point::Point()" << std::endl;
    // передаём инициализацию объекта другому методу
    read();
}

Point::Point(double x, double y)
{
    // если конструктор написать так, то для каждого скрытого поля
    // сначала вызовется КОНСТРУКТОР ПО УМОЛЧАНИЮ (для double = 0.0)
    // и только затем произойдёт присваивание
    std::cout << ">Point::Point(" << x << ", " << y << ")" << std::endl;
    init(x, y);
}

Point::Point(const Point& point)
{
    std::cout << ">Point::Point(&" << point.toString() << ")" << std::endl;
    init(point.getX(), point.getY());
}

Point::~Point()
{
    std::cout << ">Point::~Point()" << std::endl;
}

void Point::init(double x, double y)
{
    std::cout << ">Point::init(" << x << ", " << y << ")" << std::endl;
    // код ниже ДОЛЖЕН инициализировать все закрытые поля объекта
    setX(x);
    setY(y);
}

void Point::setX(double x)
{
    std::cout << ">Point::setX(" << x << ")" << std::endl;
    m_x = x;
}

void Point::setY(double y)
{
    std::cout << ">Point::setY(" << y << ")" << std::endl;
    m_y = y;
}

double Point::getX() const
{
    std::cout << ">Point::getX()" << std::endl;
    return m_x;
}

double Point::getY() const
{
    std::cout << ">Point::getY()" << std::endl;
    return m_y;
}

void Point::display() const
{
    std::cout << ">Point::display()" << std::endl;
    std::cout << toString() << std::endl;
}

std::string Point::toString() const
{
    std::cout << ">Point::toString()" << std::endl;
    std::ostringstream os;
    os << "Point(" << getX() << ", " << getY() << ")";
    return os.str();
}

void Point::read()
{
    std::cout << ">Point::read()" << std::endl;
    double x, y;
    std::cout << "Введите x: ";
    std::cin >> x;
    std::cout << "Введите y: ";
    std::cin >> y;
    init(x, y);
}

File: 00-java/src/Point.java
import java.util.Scanner;

/**
 * 2D точка
 */
public class Point {
//    /**
//     * Конструктор по умолчанию
//     */
//    public Point() {
//        System.err.println(">Point()");
//        setX(0.0);
//        setY(0.0);
//    }

    /**
     * Конструктор с параметрами
     *
     * @param x Координата X
     * @param y Координата Y
     */
    public Point(double x, double y) {
        System.err.println(">Point(" + x + ", " + y + ")");
        setX(x);
        setY(y);
    }

    /**
     * Получить текстовое представление объекта
     *
     * @return Текстовое представление объекта
     */
    @Override
    public String toString() {
        return "Point(" + getX() + ", " + getY() + ")";
    }

    /**
     * Инициализатор объекта из стандартного потока ввода
     *
     * @return Сконструированный объект из данных
     */
    public static Point read() {
        System.err.println(">Point.read()");
        Scanner scan = new Scanner(System.in);
        System.out.print("Введите x: ");
        double x = scan.nextDouble();
        System.out.print("Введите y: ");
        double y = scan.nextDouble();
        return new Point(x, y);
    }

    public Point minus(Point rhs) {
        return new Point(getX() - rhs.getX(), getY() - rhs.getY());
    }

    public double getX() {
        System.err.println(">Point.getX()");
        return m_x;
    }

    public double getY() {
        System.err.println(">Point.getY()");
        return m_y;
    }

    public void setX(double x) {
        System.err.println(">Point.setX(" + x + ")");
        m_x = x;
    }

    public void setY(double y) {
        System.err.println(">Point.setY(" + y + ")");
        m_y = y;
    }

    /**
     * Координата X
     */
    private double m_x;

    /**
     * Координата Y
     */
    private double m_y;
}
